[Step2] 문자열 계산기 PR 요청드립니다.

dajeong opened the pull request:
형준님 안녕하세요!
2단계 PR이 좀 늦었습니다..
객체지향을 어느정도까지 가져가야할지 감이 잘 안오네요 ㅎㅎ
피드백 잘 부탁드립니다. 🙏
감사합니다 🐰

dajeong committed:
[Step 2] 문자열 계산기 문제 내용 readme 작성

문제를 쉽게 참고할 수 있게 readme 파일에 작성

dajeong committed:
[Step 2] 문자열 계산기 테스트 작성

구현 전 문자열 계산기 요구사항에 맞는 테스트 작성

dajeong committed:
[Step 2] 문자열 계산기 기능 구현 및 테스트 보충

- 사칙 연산 기능 구현
- 다양한 상황의 유효성 검증 코드 추가
- 추가된 코드에 대한 테스트 작성

dajeong committed:
[Step 2] 변수명 수정

좀 더 의미있는 이름으로 수정

dajeong committed:
[Step 2] 잘못된 클래스 위치 수정

dajeong committed:
[Step 2] 프로그램 실행 클래스 생성

- 메인 메서드에서 애플리케이션을 호출하도록 구현

dajeong committed:
[Step 2] 사칙연산 테스트 하나의 테스트로 구현

- 연산 마다 생성했던 테스트 메서드를 parameterized 테스트 하나로 합치기

Flamme1004K commented on src/main/kotlin/step2/README.md:
```diff
@@ -0,0 +1,58 @@
+## 기능 요구 사항
+- 사용자가 입력한 문자열 값에 따라 사칙 연산을 수행할 수 있는 계산기를 구현해야 한다.
+- 문자열 계산기는 사칙 연산의 계산 우선순위가 아닌 입력 값에 따라 계산 순서가 결정된다. 즉, 수학에서는 곱셈, 나눗셈이 덧셈, 뺄셈 보다 먼저 계산해야 하지만 이를 무시한다.
+- 예를 들어 "2 + 3 * 4 / 2"와 같은 문자열을 입력할 경우 2 + 3 * 4 / 2 실행 결과인 10을 출력해야 한다.
```
요구 사항 정리 좋습니다 👍👍

Flamme1004K commented on src/main/kotlin/step2/Application.kt:
```diff
@@ -0,0 +1,27 @@
+package step2
```
하나 요청드릴 게 있습니다! Step으로 나누시지 말고, 모듈별로 나눠주세요. 😊

왜냐하면, 추후 step3, step4, step5로 갈수록 코드가 새로 생기기 때문에 코드가 변화하는 과정을 보고 피드백이 힘든 점이 있습니다.🙏

Flamme1004K commented on src/main/kotlin/step2/Application.kt:
```diff
@@ -0,0 +1,27 @@
+import step2.operator.*
+
+class Application {
+    private val stopCommand = "0"
```
상수에 대해서는 companion object에 정의 부탁드려요! 그리고 상수는 대문자가 컨벤션입니다 😊

링크
- [companion object](https://pearlluck.tistory.com/722)

Flamme1004K commented on src/main/kotlin/step2/Application.kt:
```diff
@@ -0,0 +1,27 @@
+                break;
+            }
+            println(INPUT_GUIDE)
+            val calculateValues = readLine()
```
kotlin 1.6의 readln()을 사용하면, input에 대한 null 체크 로직을 개선할 수 있어 보입니다.

참고 
- [kotlin readln](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/readln.html)
- [kotlin 1.6.0 released](https://blog.jetbrains.com/ko/kotlin/2021/12/kotlin-1-6-0-is-released/)

Flamme1004K commented on src/main/kotlin/step2/Application.kt:
```diff
@@ -0,0 +1,27 @@
+        println(CLOSE_GUIDE)
+        return
+    }
+}
```
Lint 확인 부탁드립니다

Flamme1004K commented on src/main/kotlin/step2/GuideScript.kt:
```diff
@@ -0,0 +1,15 @@
+
+const val RESULT_GUIDE = "# 결과 = "
+
+const val CLOSE_GUIDE = "# 종료합니다."
```
하나의 코틀린 파일에 여러 상수를 넣는다면, 모든 클래스에 대해서 사용을 할 수 있을 것 같아요. 

object class를 통하여 한번 감싸주면 어떨까요?

Flamme1004K commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,52 @@
+        return result
+    }
+
+    companion object StringValidator {
```
`StringValidator`에 대한 부분은 util 패키지로 나누는게 좋지 않을까요?

Flamme1004K commented on src/main/kotlin/step2/operator/OperatorSelector.kt:
```diff
@@ -0,0 +1,30 @@
+package step2.operator
+
+class OperatorSelector {
```
OperatorSelector에 다른 사칙연산이 추가하면 많은 클래스 변경이 필요해보여요. 어떻게 해야 최소한으로 변경할 수 있을까요?

Flamme1004K commented on src/test/kotlin/step2/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,67 @@
+import io.kotest.matchers.shouldBe
+import step2.operator.OperatorSelector
+
+class StringCalculatorTest : StringSpec({
```
꼼꼼한 테스트 너무 좋습니다! 그리고 Kotest 사용 너무 좋아요!

Flamme1004K reviewed (changes requested):
안녕하세요! 다정님!

전체적으로 꼼꼼히 구현을 해주셨군요~ 콘솔까지 구현하실 줄은 몰랐습니다! 하나의 프로그램을 만드시는 모습 너무 좋습니다.👍

몇 가지 코멘트를 남겼으니 확인 부탁드리겠습니다!

확인 후 다시 리뷰 요청 드리며, 궁금하신 점에 대해서는 DM 부탁드립니다!

마지막으로 미션 화이팅입니다! 🙌

Flamme1004K commented:
> 객체지향을 어느정도까지 가져가야할지 감이 잘 안오네요 ㅎㅎ

에 대해서 저의 생각은 이렇습니다.

객체지향을 처음부터 가져가기는 어렵다고 생각해요. 그렇기 때문에 저는 처음에는 크게 추상화를 하고, 구현을 하면서 너무 커지면 그때 역할을 나누고 있습니다.

요번 계산기를 예를 들자면, 일단 문자열 계산기 클래스를 만들고, 이 문자열 계산기에 사칙연산으로 인하여 코드가 많아지면 사칙연산에 관련된 클래스로 빼면서 서서히 객체들 간에 역할을 분담하는 쪽을 선택합니다.

아래 글들이 도움이 되셨으면 합니다.

[객체 지향 프로그래밍과 메시징](https://homoefficio.github.io/2019/07/04/%EA%B0%9D%EC%B2%B4-%EC%A7%80%ED%96%A5-%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%B0%8D%EA%B3%BC-%EB%A9%94%EC%8B%9C%EC%A7%95/)
[창시자 앨런 케이가 말하는, 객체 지향 프로그래밍의 본질](https://velog.io/@eddy_song/alan-kay-OOP)

dajeong commented:
형준님 좋은 글 공유에 꼼꼼한 리뷰 감사합니다! 🙇 
제가 깃헙은 익숙치 않아서 그러는데 다시 리뷰 요청드리려면 요고는 close하면 되는걸까요?

dajeong committed:
[Step 3] 리뷰 반영

- 정적 변수들 정리
- 패키지 명 변경 및 구조 개선
- 연산자 추가될 때 변경이 많이 일어나지 않도록 수정
- lint 에서 발견된 컨벤션 오류 수정

dajeong commented:
아 재요청이 있네요!

Flamme1004K commented on src/main/kotlin/stringcalculator/application/Application.kt:
```diff
@@ -0,0 +1,32 @@
+        OperatorSelector(OperatorRegister.operators)
+    )
+
+    companion object {
+        private const val STOP_COMMAND = "0"
+    }
```
kotlin class coding convention은 아래와 같은 순서로 이루어져 있습니다.

**1. Property declarations and initializer blocks**

**2. Secondary constructors**

**3. Method declarations**

**4. Companion object**

위의 companion object는 마지막으로 가야됩니다. 😁

참고 :

- [[jetbrain Kotlin Coding convention](https://kotlinlang.org/docs/coding-conventions.html#class-layout)](https://kotlinlang.org/docs/coding-conventions.html#class-layout)
- [[Android Kotlin 스타일 가이드](https://developer.android.com/kotlin/style-guide?hl=ko)](https://developer.android.com/kotlin/style-guide?hl=ko)

Flamme1004K commented on src/main/kotlin/stringcalculator/application/GuideScript.kt:
```diff
@@ -0,0 +1,18 @@
+package stringcalculator.application
+
+object GuideScript {
```
object 클래스를 이용한 상수 관리 좋습니다 👍

Flamme1004K commented on src/main/kotlin/stringcalculator/application/StringCalculator.kt:
```diff
@@ -0,0 +1,39 @@
+class StringCalculator(
+    private val operatorSelector: OperatorSelector
+) {
+    fun calculate(string: String?): Double {
```
`readln` 을 사용한다면, nullable을 제거할 수 있었지 않았을까요?

Flamme1004K commented on src/main/kotlin/stringcalculator/application/StringCalculator.kt:
```diff
@@ -0,0 +1,39 @@
+
+    companion object ArgumentsValidator {
+
+        fun validateLength(split: List<String>) {
+            if (split.size < 3) {
+                throw IllegalArgumentException("계산할 숫자와 연산자가 충분하지 않습니다.")
+            }
+            if (split.size % 2 == 0) {
+                throw IllegalArgumentException("숫자 또는 연산자가 누락되었습니다.")
+            }
+        }
```
validateLength 함수는 도메인 로직인 것 같은데 companion object에 넣은 이유를 알 수 있을까요?

Flamme1004K commented on src/main/kotlin/stringcalculator/operator/OperatorRegister.kt:
```diff
@@ -0,0 +1,10 @@
+package stringcalculator.operator
+
+object OperatorRegister {
```
`OperatorRegister`를 이용한 방법 좋습니다! 

enum + function을 이용하여 한번 더 리팩토링 해볼까요!? 😊 소감도 알려주세요!

링크
- [Enum 활용사례 3가지](https://jojoldu.tistory.com/137)

Flamme1004K commented on src/main/kotlin/stringcalculator/util/StringValidator.kt:
```diff
@@ -0,0 +1,18 @@
+
+object StringValidator {
+    fun validateNumber(targetNumber: String) {
+        if (!targetNumber.matches(Regex("[0-9]+"))) {
```
`Strings.matches` 는 생성 비용이 비싼 객체에요.

이를 해결하려면 어떻게 해야할까요? 🤔

Flamme1004K reviewed (approved):
안녕하세요! 다정님!

피드백 반영하시느라 고생하셨습니다! 👍👍

다음 단계를 위하여 머지하도록 하겠습니다!🙌

몇 가지 코멘트를 남겼으니, 다음 미션 때 함께 진행해 주시면 되겠습니다! 😆

궁금하신 점에 대해서는 DM 부탁드리겠습니다. 🙇🏻

다음 미션도 화이팅입니다! 🤩
